### PR TITLE
Accept `Any` filter for `find_one` methods

### PR DESCRIPTION
**Pull request by armanddidierjean**
According to motor and motor-asyncio, we can provide either 
 - a mapping like `{str: Any}`
 - any other object, used as `{"_id": Any}`

> filter (optional): a dictionary specifying the query to be performed OR any other type to be used as the value for a query for "_id".
> - [`find_one` documentation](https://motor.readthedocs.io/en/stable/api-asyncio/asyncio_motor_collection.html#motor.motor_asyncio.AsyncIOMotorCollection.find_one)

We could either use the type `typing.Optional[typing.Union[typing.Mapping[str, typing.Any], typing.Any]]` or `typing.Optional[typing.Any]`. If the second syntax is more concise and simple, the first one put an emphasis on the two possibilities: the mapping or any other type.

It does not seem possible to use an Any type for `find_one_and_delete`, `find_one_and_replace` and `find_one_and_update`

---

As pointed by @L0RD-ZER0, and unlike [`find` documentation](https://motor.readthedocs.io/en/stable/api-asyncio/asyncio_motor_collection.html#motor.motor_asyncio.AsyncIOMotorCollection.find), `find` fail at runtime if we don't provide a mapping.

### CODE REVIEW

**Comment by L0RD-ZER0** [comment]
> According to motor and motor-asyncio, we can provide either 
>  - a mapping like `{str: Any}`
>  - any other object, used as `{"_id": Any}`
> 
> > filter (optional): a dictionary specifying the query to be performed OR any other type to be used as the value for a query for "_id".

If you were to check the implementation of ``find`` or ``find_one`` in motor, (yes, I did go over the implementation of this and pymongo internals before writing the stubs,) you'll see that ``find_one`` is delegated to pymongo's implementation and ``find`` creates a ``pymongo.Cursor`` instance, which is further wrapped in ``AgnosticCursor``.

As per the implementation of ``pymongo.Cursor``, the only valid types that it can accept are the ones which behave similar to/inherit the behavioural patterns from ``typing.Mapping[str, Any]``, which is appropriately typed there. If we add an ``Any`` type to the argument, in addition to the ``Mapping[...]``, it would be misleading, and might prompt something like ``find("this-would-work-with-Any")`` and we won't get a linter or type checker error, whereas it would break down at runtime.

> "... OR any other type to be used as the value for a query for "_id"."

As for this portion, after going through the implementation, I deducd that whosoever documented that portion had the intent of conveying "you can mention the value of identifier (`_id`) field as well," which is the thought I'm working upon. So at the end of the day, it would be a mapping if present and typing it as ``Any`` wouldn't be good for type analysis. 

**TLDR**; it all boils down to what ``filter`` type ``pymongo.Cursor`` accepts, which happens to be ``Mapping[str, Any]`` and that's why it's typed there.

If you have any other questions, comment on this issue. If nothing else, I'll be closing it in a few hours.

Thanks for the PR! Have a great day!

**Comment by armanddidierjean** [comment]
Hi,

Thanks for the detailed answer and the various explanations.

I'm honestly not completely sure that an `Any` object would not be accepted but I may be mistaken something.

If this is the right `find_one` [pymongo's implementation](https://github.com/mongodb/mongo-python-driver/blob/0d44783eddec96a9ae1a03b148cb4f049dca3e49/pymongo/collection.py#L1477), I have the impression that the method would accept a parameter with a type Any: `filter: Optional[Any]`.
Then, if `filter` is not an instance of `typing.Mapping` a transformation is done:
```python
if filter is not None and not isinstance(filter, abc.Mapping):
    filter = {"_id": filter}
```
Finally, pymongo instantiate a `pymongo.Cursor`, which as you said can only use a `Mapping[str, Any]`.

I'm not completely sure that motor use pymongo `find_one` method, it may create a `pymongo.Cursor` instance directly. I can confirm that the syntax `db.find_one(ObjectId("myid"))` works at runtime.

Thanks for this library which is definitely life changing when working with motor.

**Comment by L0RD-ZER0** [comment]
> I'm not completely sure that motor use pymongo `find_one` method, it may create a `pymongo.Cursor` instance directly. I can confirm that the syntax `db.find_one(ObjectId("myid"))` works at runtime.

This one's interesting. So it works with ``ObjectId`` but not with ``str`` or ``int``? I'll look into this one again. Thanks for pointing that out.

> Thanks for this library which is definitely life changing when working with motor.

Your support for the project is quite appreciated! I am hoping this lib can get a bunch of docstrings added soon so it becomes even better and moves to a proper v1.0 release. (something on local branch I work on time to time, halfway-ish done)

**Comment by L0RD-ZER0** [comment]
> > I'm not completely sure that motor use pymongo `find_one` method, it may create a `pymongo.Cursor` instance directly. I can confirm that the syntax `db.find_one(ObjectId("myid"))` works at runtime.
> 
> This one's interesting. So it works with ``ObjectId`` but not with ``str`` or ``int``? I'll look into this one again. Thanks for pointing that out.

With regards to this, ``Collection.find_one(Any)`` indeed works. As you said, it converts `<Obj>` to `{'_id': <Obj>}` if it's not ``None`` or ``Mapping``. You are correct there. This one's an oversight in my part it seems.

However the same can not be said for ``Collection.find(...)``. It directly creates an instance of the cursor. There is no sort of filter conversion for this one.

Therefore, ``find_one`` can accept ``Any``, whereas ``find`` only accepts a ``Mapping[str, Any]``.

I'll request you to make the change, and I'll review and merge the PR tomorrow if you don't mind. Thanks for pointing this one out! Your help is much appreciated!

**Review by L0RD-ZER0 — CHANGES_REQUESTED** [review]
I'll request you to revert the types on ``find`` and the types in ``find_one`` are alright. I'll take your advice from before and would go with ``Union[Mapping[...], Any]`` for the sake of clarity as of now. If it would be changed in future is something our future selves would have to deal with.

**Comment by armanddidierjean** [comment]
You are completely right, unlike `find_one`, `find` requires a Mapping at runtime. I falsely assumed otherwise after reading motor documentation. 

I rebased the branch to remove typing changes for `find_one`, let me know if other changes are needed.

**Comment by L0RD-ZER0** [comment]
It should be good now. Let me review and merge real quick. Thanks for your contribution!!

**Comment by armanddidierjean** [comment]
Your welcome, the pleasure was mine.

**Review by L0RD-ZER0 — APPROVED** [review]
Alright, this looks good. I'll proceed with the merge.
